Contribute: Rewrite and Restructure (#123)
mumble-voip/mumble-www: hugo/content/contribute.md

### Additional projects

#### 3rd party Interfaces

Web interfaces, web-viewers, administration programs, CMS website modules – all neat things we would love to see! These could be written in any language supported by our Ice- or gRPC-based RPC Interfaces. You could start your own project or link up with an {{< wiki "3rd Party Applications" "already existing one" />}}. To get an idea about how interaction with the Mumble server might look like, you can take a look at the example scripts in our [GitHub repository for mumble-scripts](https://github.com/mumble-voip/mumble-scripts).

#### Ice authenticators

See {{< wiki Authenticators />}}. We’re happy for every authenticator and authenticator improvement.

#### Positional Audio plugins

Most of our [positional audio plugins](https://wiki.mumble.info/wiki/Games#Positional_audio) regularly break on game updates. To read on how you can help us update the plugins see the {{< wiki Pluginguide />}} page. 

> You can also add positional audio support to Games via our standard `Link` plugin. This way it will not get outdated, but you need permission to change (and access to) the Source Code of a game though (e.g. via an open license or developer permission). See {{< wiki "Link" "Linking a game to Mumble" />}}.

Author: @Krzmbrzl Check again please.

Reviewer: I would remove the "though" but otherwise this LGTM

Author: Ok, I will remove the "though".

toc: true
---

You can contribute to the Mumble Project in multiple ways:

<!-- Table of Content -->

- [Report Bugs and suggest Improvements/Features](#report-bugs-and-feature-requests)
- [Help with Development](#development)
    - [Coding](#coding)
        - [Additional projects](#additional-projects)
    - Non-Coding: 
        - [Translation](#translation)
        - [Testing](#testing)
        - [GUI/Mumble Themes](#mumble-theme)
- Improve the [Documentation](#documentation) and [Website](#website)
- [Support other users & take part in our community](#community)
- [Host a Mumble server](#hosting)
- [Promote Mumble](#promotion)

---

# Report bugs and Feature Requests

1. Take a look at existing [issues in our GitHub repo](https://github.com/mumble-voip/mumble/issues) to check whether your issue or request was already reported.

>    If that's the case, then vote for it with a thumps up or add useful information & ideas with a comment.

Reviewer: ```suggestion
   If that's the case, then vote for it with a thumbs up or add useful information & ideas with a comment.
```

   
2. If there is no issue report yet, then create a [new one](https://github.com/mumble-voip/mumble/issues/new/choose).

**Note:** For security issues, see [Report a vulnerability]({{< relref "security#reporting-a-vulnerability" >}}).

# Development

We are always looking for helping hands willing to contribute to Mumble.

## Coding

Mumble is mostly written in **C++** using the **Qt** library.
Thus previous knowledge in C++ and Qt is definitely useful, but we also encourage interested people who want to learn it to give it a try. We're happy to help you out, if needed.

### Getting Started

> The best way to get into Mumble development is to link up with us on IRC ([#mumble-dev on freenode](irc://chat.freenode.net/mumble-dev)), also available via Matrix (Join: `#freenode_#mumble-dev:matrix.org` or use [Direct Link via Element.io (Web App)](#freenode_#mumble-dev:matrix.org)).

Reviewer: Does clicking the matrix link actually work? 'cause it seems like it would not :thinking:

Author: Somehow it's the wrong link...will fix.

Alternatively you can create or comment (on) a pull request or issue report on GitHub.

For more information, see the categories below:

#### Source Code

Our source code is hosted on GitHub:

- [Overview of Mumble Projects/Repos](https://github.com/mumble-voip/)

- [Main Mumble Repo](https://github.com/mumble-voip/mumble): includes the Server and Client
  
- [iOS app](https://github.com/mumble-voip/mumble-iphoneos): unmaintained; help wanted

#### Documentation regarding Development

Take a look at the following documentation:

- [Contributing guidelines](https://github.com/mumble-voip/mumble#contributing)

- [Development Documentation](https://github.com/mumble-voip/mumble/tree/master/docs/dev)

> - [Building Instructions](https://github.com/mumble-voip/mumble#building)

Reviewer: ```suggestion
- [Build Instructions](https://github.com/mumble-voip/mumble#building)
```

We are thankful for any contribution on translating. Please check the {{< wiki "Language Translation" "Language Translation wiki page" />}} for more information.

### Testing

To find bugs and polish new releases we need testers.

If you want to test the newest development versions, you have two options:

- We provide development snapshots. You can download them from our [Downloads page]({{< relref "downloads#development-snapshots" >}}).

- Build the latest development version of Mumble, see [Building Mumble](https://github.com/mumble-voip/mumble#building).

If you find any issues, take a look at [Report Bugs](#report-bugs-and-feature-requests).

### Development Documentation

See [Section "Documentation" below](#documentation). 

### Mumble theme

You find the official Mumble theme in its own [repository on GitHub](https://github.com/mumble-voip/mumble-theme).

#### Additional themes

Additional Mumble themes bring variety into Mumble's design and choices for different tastes.

> Create new themes for Mumble! Information can be found on our {{< wiki Skinning "Skinning wiki page" />}} and existing skins we know about on the {{< wiki Skins "Skins wiki page" />}}.

Reviewer: ```suggestion
Create new themes for Mumble! Information can be found on our {{< wiki Skinning "Skinning wiki page" />}} and existing themes we know about on the {{< wiki Skins "Skins wiki page" />}}.
```

# Documentation & Website

Improve our Documentation and Website.

## Documentation

Right now you find most of the **(User) Documentation** in our [wiki](https://wiki.mumble.info).

Any contributions are welcome: improving existing pages or creating new ones with useful content. Also see the {{< wiki "Category:Please_Improve" "Please Improve Category" />}}.

You can create a wiki account {{< wiki "Special:RequestAccount" "here" />}}.

> The **Development Documentation** is now in [Mumble's GitHub repo](https://github.com/mumble-voip/mumble/tree/master/docs).

Reviewer: ```suggestion
The **Development Documentation** is now in [Mumble's GitHub repo](https://github.com/mumble-voip/mumble/tree/master/docs/dev).
```

- [Host a Mumble server](#hosting)
- [Promote Mumble](#promotion)

---

# Report bugs and Feature Requests

1. Take a look at existing [issues on GitHub](https://github.com/mumble-voip/mumble/issues) to check whether your issue or request was already reported.
   If that's the case, then vote for it with a thumps up or add useful information & ideas with a comment.
   
2. If there is no issue report yet, then create a [new one](https://github.com/mumble-voip/mumble/issues/new/choose).

# Development

## Coding

We are always looking for helping hands willing to contribute to Mumble.

Mumble is mostly written in **C++** using the **Qt** library.
Thus previous knowledge in C++ and Qt is definitely useful, but we also encourage interested people who want to learn it to give it a try. We're happy to help you out, if needed.

Our source code is hosted on GitHub:

- [Overview of Mumble Projects/Repos](https://github.com/mumble-voip/)

- [Main Mumble Repo](https://github.com/mumble-voip/mumble): includes the Server and Client
  
- [iOS app](https://github.com/mumble-voip/mumble-iphoneos): unmaintained; help wanted

### Getting Started

> The best way to get into Mumble development is to link up with us on IRC ([#mumble-dev on freenode](irc://chat.freenode.net/mumble-dev)), also available via Matrix (Join: `#freenode_#mumble-dev:matrix.org` or use [Direct Link via Element.io (Web App)](#freenode_#mumble-dev:matrix.org)).

Author: Could someone test the Matrix method with:
`#freenode_#mumble-dev:matrix.org` ?

Reviewer: Using `/join #freenode_#mumble-dev:matrix.org` seems to work

Author: Very good, then I assume this as resolved.

Alternatively you can create or comment on a Pull request or issue.

### Documentation regarding Development

- [Contributing guidelines](https://github.com/mumble-voip/mumble#contributing)

- [Development Documentation on Github](https://github.com/mumble-voip/mumble/tree/master/docs/dev)

- [Building Instructions](https://github.com/mumble-voip/mumble#building)

<!-- Should be implemented once it's ready:
- (Work-in-Progress) [C++ guidelines & tools](https://github.com/mumble-voip/mumble/discussions/4261) 

-->

### Open issues and ongoing development

Take a look at [issues](https://github.com/mumble-voip/mumble/issues) and [pull requests](https://github.com/mumble-voip/mumble/pulls) on GitHub, especially with the following Labels:

> <!-- add label for important tasks? -->

Author: @Krzmbrzl Could you maybe also ask that in one your meetings?

Reviewer: Will do

Reviewer: Okay so the idea will be that we'll assess new bug reports on how critical we think they are. If they are major, we'll label them has high prio and we want to try to fix them before a new release.
This will apply to bugs only though.

Author: Ok, should we mention that label in this part then?

Reviewer: Yeah we can mention it.
We did not really conclude on which label it is exactly that we want to use for that... :thinking:

Author: Maybe just "important" ?

Reviewer: We already have the different `Prio` labels so it would probably make sense to reuse them. But given that we have different priority levels, this contradicts the idea of having a binary choice for this :thinking: 

Maybe it is time to delete the Prio labels and create a new `important` label instead. We would have to look at the issues that currently carry a prio label in order to see if any of those should have the new label instead...

Author: I will wait then for your decision.

By the way: Should I add the `bug` label as well?
https://github.com/mumble-voip/mumble/labels/bug

Reviewer: Yeah adding the `bug` and `feature-request` label might be a good thing here :+1:

Author: added this to the todo list, so i resolve the conversation.

### Additional projects

#### 3rd party Interfaces

Web interfaces, web-viewers, administration programs, CMS website modules – all neat things we would love to see! These could be written in any language supported by our Ice based or gRPC based RPC Interfaces. You could start your own project or link up with an {{< wiki "3rd Party Applications" "already existing one" />}}. To get an idea about how interaction with the Mumble server might look like, you can take a look at the example scripts in our [GitHub repository script folder](https://github.com/mumble-voip/mumble/tree/master/scripts).

#### Ice authenticators

See {{< wiki Authenticators />}}. We’re happy for every authenticator and authenticator improvement.

#### Positional Audio plugins

Most of our positional audio plugins regularly break on game updates. To read on how you can help us update the plugins see the {{< wiki Pluginguide />}} page. 

> You can also add native positional audio support to an Open Source game (or a game where you are allowed to edit the source code (for example if you are a developer)), see {{< wiki "Link" "Linking a game to Mumble" />}}.

Author: Is this correct? @Kissaki

Reviewer: That you can use the Link plugin for that? Yes that is correct

Author: :thinking: To be more precise: So the formulation of the second sentence is correct?
- add native positional audio support to a game with that Link code?
- and how does that work then; without any plugin?

Reviewer: `Link` is actually the plugin that is used for that. It works by providing an area of shared memory into which some endpoint on the game's size continuously writes the position information. And the Link plugin then fetches this information from the shared memory in order to send it to Mumble.

Author: Ok, thanks for the info. So there is no need for additional plugins in this case?

    - [Coding](#coding)
        - [Additional projects](#additional-projects)
    - Non-Coding: 
        - [Translation](#translation)
        - [Testing](#testing)
        - [GUI/Mumble Theme](#mumble-theme) and [Skins](#skins)
- Contribute to the [Documentation](#documentation) and [Website](#website)
- [Support other users & take part in our community](#community)
- [Host a Mumble server](#hosting)
- [Promote Mumble](#promotion)

---

<!-- Content Categories -->

# Report bugs and Feature Requests

1. Take a look at existing [issues on GitHub](https://github.com/mumble-voip/mumble/issues) to check whether your issue or request was already reported.
   If that's the case, then vote for it with a thumps up or add useful information & ideas with a comment.
   
2. If there is no issue report yet, then create a [new one](https://github.com/mumble-voip/mumble/issues/new/choose).

<!-- Also take a look at Testing <!-- add link to section below --> 

# Development

## Coding

<!-- Todo: 
- mention Label for important tasks ? (e.g. [gRPC development](https://github.com/mumble-voip/mumble/pull/3947))

-->

We are always looking for helping hands willing to contribute to Mumble.

> Mumble is mostly written in **C++** and uses many **Qt** libraries.

Author: @Krzmbrzl  Is that ok?

> So knowledge in C++ and Qt is useful, but we also accept interested people you want to learn.

Author: @Krzmbrzl Is that ok?

> Mumble is mostly written in **C++** and uses many **Qt** libraries.
> So knowledge in C++ and Qt is useful, but we also accept interested people you want to learn.

Reviewer: ```suggestion
Mumble is mostly written in **C++** using the **Qt** library.
Thus previous knowledge in C++ and Qt is definitely useful, but we also encourage interested people who want to learn it to give it a try. We're happy to help you out, if needed.
```

   If that's the case, then vote for it with a thumps up or add useful information & ideas with a comment.
   
2. If there is no issue report yet, then create a [new one](https://github.com/mumble-voip/mumble/issues/new/choose).

# Development

## Coding

<!-- Todo: 
- mention Label for important tasks ? (e.g. [gRPC development](https://github.com/mumble-voip/mumble/pull/3947))

-->

We are always looking for helping hands willing to contribute to Mumble.

Mumble is mostly written in **C++** using the **Qt** library.
Thus previous knowledge in C++ and Qt is definitely useful, but we also encourage interested people who want to learn it to give it a try. We're happy to help you out, if needed.

Our source code is hosted on GitHub:

- [Overview of Mumble Projects](https://github.com/mumble-voip/)

- [Main Mumble Repo](https://github.com/mumble-voip/mumble): includes the Server and Client
  
- [iOS app](https://github.com/mumble-voip/mumble-iphoneos): unmaintained; help wanted

### Getting Started

The best way to get into Mumble development is to link up with us on IRC ([#mumble-dev on freenode](irc://chat.freenode.net/mumble-dev)).

> <!-- is this still the best method to get into contact? -->

Reviewer: You mean for bug reports and feature/change requests? Yes it is. Except for reporting security related issues in which case it'd be best to report them via mail to `security@mumble.info`

For further questions we should refer to the forums and as a last resort, we could mention `contact@mumble.info`

> <!-- is this still the best method to get into contact? -->

Author: This is not the part that refers to.

If you look at the whole document, you see that this is about contacting you for development over IRC.

> <!-- is this still the best method to get into contact? -->

Reviewer: Ah okay my bad. Yes IRC (#mumble-dev channel on Freenode or via Matrix: https://app.element.io/#/room/!VNUpYnUPdhTAqagvUu:matrix.org) is still the way to go. Additionally creating a PR or issue (or commenting on an existing one) would also work just fine :point_up:

> <!-- is this still the best method to get into contact? -->

Author: Is there some generic link for your matrix channel/server?
I only found links that include web applications, like your link above.
(Sry if I maybe missunderstand something, I don't use matrix)

> <!-- is this still the best method to get into contact? -->

Reviewer: I have not yet found a way to link to IRC channels and the matrix link is the only one that I know.
For everyone that wants to use plain IRC we simply have to tell them to connect to Freenode and then join the #mumble-dev channel

> <!-- is this still the best method to get into contact? -->

Author: For IRC this link was already included in the original file: irc://chat.freenode.net/mumble-dev
(the browser will then ask for an application to open the link)
I suspect that something similar might exist for matrix.

> <!-- is this still the best method to get into contact? -->

Author: It seems it could be this: `#freenode_#mumble-dev:matrix.org`   (copied from server search)

Examples given for normal connections in many matrix clients is: `Type /join #matrix:matrix.org`

> <!-- is this still the best method to get into contact? -->

Reviewer: Idk. Could work, yes.
The link I provided lets you choose which application you want to use to open it though (admittedly it is a bit hidden, but there is a "change" button that lets you choose another app) :shrug:

> <!-- is this still the best method to get into contact? -->

Author: I will add both.

### Documentation regarding Development

- [Contributing guidelines](https://github.com/mumble-voip/mumble#contributing)

<!-- Should be implemented once it's ready:
- (Work-in-Progress) [C++ guidelines & tools](https://github.com/mumble-voip/mumble/discussions/4261) 

-->

> <!-- add developer documentation -->

Reviewer: We could link to https://github.com/mumble-voip/mumble/tree/master/docs/dev in general and to https://github.com/mumble-voip/mumble/tree/master/docs/dev/build-instructions

#### Other 3rd party projects

If you are interested in more projects (Clients, Servers, cli tools etc.), you can also take a look at {{< wiki 3rd_Party_Applications />}}.
<!-- https://wiki.mumble.info/wiki/3rd_Party_Applications -->
Many of these projects will either gladly accept help or can (often) be forked easily, because of the open licenses.

## Non-Coding

### Translation

Mumble currently (April 2020) supports over 15 different languages (38 if including unfinished translations). As more people get to know and use Mumble, versions in their native languages become more and more important.

We are thankful for any contribution on translating. Please check the {{< wiki "Language Translation" "Language Translation wiki page" />}} for more information.

### Testing

To find bugs and polish new releases (before and after release) we need testers.

If you want to test the newest development versions, you have two options:

- We provide development snapshots. You can download them from our [Downloads page]({{< relref "downloads" >}}).

> <!-- link to exact point? https://www.mumble.info/downloads/#development-snapshots -->

Reviewer: Yep: https://www.mumble.info/downloads/#development-snapshots

On the website you can hover over headlines and click the anchor that appears to its left in order to get a link to that headline :point_up:

> <!-- link to exact point? https://www.mumble.info/downloads/#development-snapshots -->

Author: I posted the exact same link in the box visible above; the question was about this `relref` tag.

> <!-- link to exact point? https://www.mumble.info/downloads/#development-snapshots -->

Reviewer: > the question was about this relref tag.

Well then you should have made that clear :P 

I don't think that `relref` supports anchors. Could be wrong though. @Kissaki probably knows that best.

> <!-- link to exact point? https://www.mumble.info/downloads/#development-snapshots -->

Author: Seems relref supports anchors: https://gohugo.io/functions/relref/
So I will include it this way.

#### Additional themes

Additional Mumble themes bring variety into Mumble's design and choices for different tastes.

Create new themes for Mumble! Information can be found on our {{< wiki Skinning "Skinning wiki page" />}} and existing skins we know about on the {{< wiki Skins "Skins wiki page" />}}.

<!-- Wiki links should be renamed, once the naming in the wiki has changed. -->

# Documentation & Website

Improve our Documentation and Website.

## Documentation

<!-- Todo: 

- We should add guidelines for style etc.
Maybe we could put that somewhere in the github repo instead.

See also: https://github.com/mumble-voip/mumble-www/discussions/125 and https://github.com/mumble-voip/mumble/discussions/4648

Add links to issues for a concrete plan:

- add Labels for important issues and pull requests

-->

We plan to move all our documentation to the Website and GitHub Repo respectively.

Right now you find most of the Documentation in our
[wiki](https://wiki.mumble.info).

Any contributions are welcome, improving existing pages or creating new ones with useful content. Also see the {{< wiki "Category:Please_Improve" "Please Improve Category" />}}.

> The new places for documentation will probably be here:

Author: Could we have a decision about the places (seperate documentation repo e.g.) and some content ideas & structure concept before merge?
Otherwise people will not really know what to do and how to help, aside from editing the existing wiki.
This could of course also be an intermediate solution, to tell people to just edit the wiki for now.

I know we already had this discussion, but it's no solution to simply say: there will be no answer.
No answer, no content.

For example: https://github.com/mumble-voip/mumble-www/discussions/125#discussioncomment-274784

Reviewer: I will add the discussion on whether we want a separate repo or not to the list of topics for our next meeting :point_up:

Author: To summarize additional main questions:
1. Should we encourage potential contributors to create new markdown files for documentation (also by copying & updating the wiki information)?
2. Is it planned to remove the wiki enterely?
3. Is it planned to somehow automatically convert the wiki?
4. Does someone in the team want to make a (short) plan for the transition and new documentation?
So contributors have a rough idea about the structure (Categories, Topics, what should be split), potential and wanted contents, writing style etc. (also in regards to the existing [contents of the wiki](https://wiki.mumble.info/wiki/Special:AllPages)).

Reviewer: See https://github.com/mumble-voip/mumble-www/discussions/125#discussioncomment-287246

Author: Ok, so I think we now have three options for this page:
1. Only mention the wiki for now.
2. Mention the wiki, but tell users that a change is planned (stay tuned)
3. Option 2 + that people are free to already contribute something for user documentation (in this repo) and/or development documentation (in the main repo).

Reviewer: I think we should refer them to he wiki for the time being

Author: Ok.

- [Mumble GitHub repo - Docs](https://github.com/mumble-voip/mumble/tree/master/docs)

- [Website's GitHub repo - Documentation](https://github.com/mumble-voip/mumble-www/tree/master/hugo/content/documentation)

#### Documentation Issues & Pull requests

Also take a look at open Issues and Pull Requests:

- [Documentation issues in Mumble's GitHub repo](https://github.com/mumble-voip/mumble/labels/documentation)

- [Pull Request with TYPE label "Docs"](https://github.com/mumble-voip/mumble/pulls?q=is%3Apr+is%3Aopen+docs)

- [Documentation issues in Mumble's Website repo](https://github.com/mumble-voip/mumble-www/labels/docs)

---

<!-- 
**Ice methods**: 
<!-- outdated? -->

> <!-- Is this here now: https://www.mumble.info/documentation/slice/1.3.0/html/Murmur.html  ? -->

Author: Should this be included?
@Kissaki

Author: I will keep this part, but commented out.

-Link to Writing guidelines (markdown etc.), Planned Content etc. (should be in the Readme.md of the mumble-www repo)

-->

This website has a GitHub repository: [mumble-www](https://github.com/mumble-voip/mumble-www)

You can:

- report [issues](https://github.com/mumble-voip/mumble-www/issues)
- create [requests](https://github.com/mumble-voip/mumble-www/issues), 
- create pull requests for changes to the website

## Community

Be a part of our community and help others!

* Our [issue tracker](https://github.com/mumble-voip/mumble/issues)
* [Discussions on GitHub](https://github.com/mumble-voip/mumble/discussions)
* IRC: [#mumble on irc.freenode.net](irc://irc.freenode.org/mumble)
* Chat: [Mumble room on Riot/Matrix](https://riot.im/app/#/room/#mumble:matrix.org) (shares messages with IRC)
* Our [Forums](https://forums.mumble.info/)

# Hosting

Host your own Mumble Server and register it on our Public Server List.
<!-- add new Server guide (once it's ready) -->
See our {{< wiki "Running_Murmur" "Server Guide" />}} and [Server Registration details](https://wiki.mumble.info/wiki/Murmur.ini#Server_Registration).

You find a list of hosters on our wiki: {{< wiki Hosters />}}.

> <!-- Is that recent and trustworthy information? Maybe it would be better to have a list that is maintained by the team, so there can be a (little) check before an entry is added to the list -->

Author: ?

Reviewer: What kind of check would we be able to perform? We can't rent a server from every hoster just to see if and how well it works :thinking:

Author: Two things:
1. It can be a legal problem for you, if you wildly recommend server hosters without checking them at all.
2. The performance or usability is not the main problem; the main problem is whether these services are trustworthy or not. So at least a short look at the website, check for https, certificate details, check for obvious things like wrong spelling everything, should be possible.
Some countries also offer company registration sites, where you can check whether a company exists or not.

All in all, I would recommend to delete this or just recommend bigger, trustworthy services.
If people want to find more hosters, they can search for them themselves.

Reviewer: > Users might trust the information listed there, even though there is no security at all.

This is nothing that we can prevent from happening. On the contrary: if we start doing little tests here and there then some users might overestimate the value of these tests and then put more trust in these recommendations than if they had known that this is just a more or less uncurated list.

> It can be a legal problem for you, if you wildly recommend server hosters without checking them at all.

We don't recommend, we list. And we are not responsible for what these hosters do. We only have to act if we know that they are scamming people or something like that.

> the main problem is whether these services are trustworthy or not

Trustworthy can be a bit subjective as well. And even only checking for these kind of things is work that someone has to invest the time to do. Personally I don't have the time to do that as well :shrug: 

> All in all, I would recommend to delete this or just list bigger, trustworthy services.

That sounds reasonable to me. @davidebeatrici @Kissaki @ZeroAbility what's your opinion on this?

Author: I deleted the link for now.

<!-- Table of Content -->
<!-- manually created for now -->

- [Report Bugs and suggest Improvements/Features](#report-bugs-and-feature-requests)
- [Help with Development](#development)
    - [Coding](#coding)
        - [Additional projects](#additional-projects)
    - Non-Coding: 
        - [Translation](#translation)
        - [Testing](#testing)
        - [GUI/Mumble Theme](#mumble-theme) and [Skins](#skins)
- Contribute to the [Documentation](#documentation) and [Website](#website)
- [Support other users & take part in our community](#community)
- [Host a Mumble server](#hosting)
- [Promote Mumble](#promotion)

---

<!-- Content Categories -->

# Report bugs and Feature Requests

1. Take a look at existing [Issues on Github](https://github.com/mumble-voip/mumble/issues) to check whether your Issue or Request was already reported.
   If that's the case, then vote for it with a thumps up or add useful information & ideas with a comment.
   
2. If there is no Issue report yet, then create a [new one](https://github.com/mumble-voip/mumble/issues/new/choose).

<!-- Also take a look at Testing <!-- add link to section below --> 

# Development

## Coding

<!-- Todo: 

> - mention important tasks ? (e.g. [gRPC development](https://github.com/mumble-voip/mumble/pull/3947))

Author: Then how about having a label for that. I already suggested below to mention the priority labeled issues, but it is not really the same imho.

I would like to have an additional label that marks issues and pull request that are important for the general development, like the refactoring, gRPC rewrite etc..
These are maybe not urgent (like the priority labeled issues), but they are important in the long run.

> - mention important tasks ? (e.g. [gRPC development](https://github.com/mumble-voip/mumble/pull/3947))

Reviewer: Hm I think what is considered to be important is probably very subjective :thinking:

> - mention important tasks ? (e.g. [gRPC development](https://github.com/mumble-voip/mumble/pull/3947))

Author: Well, the team should decide about what they/you think is important.

And there is of course also an objective view to whats important: security, core functionality, outdated features and standards etc.

> - mention important tasks ? (e.g. [gRPC development](https://github.com/mumble-voip/mumble/pull/3947))

Reviewer: I don't think this really is worth the hassle. As I said before: People will simply work on features/things that are important to them. So I don't think that marking any issues as important will change that :shrug:

> - mention important tasks ? (e.g. [gRPC development](https://github.com/mumble-voip/mumble/pull/3947))

Author: In the past it seems it was considered worth the hassle. But if you don't want that.

-->

We are always looking for helping hands willing to contribute to mumble.

> You'll need a good understanding of **C++** and knowledge in **Qt** is definitely helpful.

Author: Well, do you really want to deal with people who have no knowledge yet?

> You'll need a good understanding of **C++** and knowledge in **Qt** is definitely helpful.

Reviewer: Believe it or not I did not really know C++ when I started working on Mumble. I taught it to myself on the way.

And if there is someone who really doesn't know shit, I think that they'll realize that pretty quickly and give up anyways. But this way someone who has a general idea about programming but maybe not specifically with C++/Qt (as I did) might still give it a shot.

I think if we state that "a good understanding of C++" is required a lot of people will not even look into this as they don't deem themselves C++ expert (maybe even if they generally know C++)

> You'll need a good understanding of **C++** and knowledge in **Qt** is definitely helpful.

Author: Ok.

> I think if we state that "a good understanding of C++" is required a lot of people will not even look into this as they don't deem themselves C++ expert (maybe even if they generally know C++)

That might be true.
In that case I would choose a completely different formulation though.
The sentence above is just copied from the old text.
I was thinking about just telling people that Mumble is mostly written in C++ and is using libraries (or whatever it's correctly called in this context) from Qt.
In addition we could add a sentence, that you are willing to also accept people who want to learn :wink: .

> You'll need a good understanding of **C++** and knowledge in **Qt** is definitely helpful.

Reviewer: Sounds good :+1:

> Our source code is hosted on Github:

Author: I don't know if it's always the right thing to use the marketing wording.
But if you insist.

> Our source code is hosted on Github:

Reviewer: Well that's what the platform is called, so I would want to stick to that :shrug:

- [Overview of Projects](https://github.com/mumble-voip/)

- [Main Mumble](https://github.com/mumble-voip/mumble) (includes the Server and Client)
  
- [iOS app](https://github.com/mumble-voip/mumble-iphoneos) (unmaintained; help wanted)

> **Getting Started**

Author: If we will not have duplication of this header, I agree.

The best way to get into mumble development is to link up with us on IRC ([#mumble-dev on freenode](irc://chat.freenode.net/mumble-dev)).

### Documentation regarding Development

> - [Contributing Guidelines](https://github.com/mumble-voip/mumble#contributing)

Author: Well, the first points to the second, so if there will be no content added to first; I chose to only use the first, as the title and folder of the second might change.

> - [Contributing Guidelines](https://github.com/mumble-voip/mumble#contributing)

Reviewer: Fair enough :+1:

> - (Work-in-Progress) [C++ Guidelines & tools](https://github.com/mumble-voip/mumble/discussions/4261) 

Author: It would be best to put the results into a file and link to that.

> - (Work-in-Progress) [C++ Guidelines & tools](https://github.com/mumble-voip/mumble/discussions/4261) 

Reviewer: Yeah - except I think we first should start integrating these practices in our code base to see what works well and what doesn't.
At this stage I am not too confident writing all of that as a coding guideline for our project (especially given that the existing code basically uses none of that xD)

> - (Work-in-Progress) [C++ Guidelines & tools](https://github.com/mumble-voip/mumble/discussions/4261) 

Author: So I will comment it out for now.

> ### Open Issues and ongoing Development

Author: Well it always feels wrong to me to use the small variant everywhere, but I guess that's the difference between languages.

> ### Open Issues and ongoing Development

Reviewer: Yeah it took me a while to get used to that as well (and yet it still happens to slip through my attention) ^^

Take a look at [Issues](https://github.com/mumble-voip/mumble/issues) and [Pullrequests](https://github.com/mumble-voip/mumble/pulls) on Github, especially with the following Labels:

- [Good first issue](https://github.com/mumble-voip/mumble/labels/good%20first%20issue)

> - [Help wanted](https://github.com/mumble-voip/mumble/labels/help%20wanted)

Author: No, I think you should fix the labels instead.
Even though as far as I have seen, most issues have the first label and the others have both...

> - [Help wanted](https://github.com/mumble-voip/mumble/labels/help%20wanted)

Author: Or do you think the labels are intentional, like `help needed` implies that no one in the team can do it?
I have two labels:
- `help wanted`
- `important help wanted`

For example.

> - [Help wanted](https://github.com/mumble-voip/mumble/labels/help%20wanted)

Reviewer: Yes to my understanding `help wanted` indicates that we consider this a good starting point for external contributors whereas `help needed` really implies that no core developer has expertise in this area and thus we are actually dependent on external help top solve the issue (properly).

Well the issue doesn't have to be _important_ in order to get the `help needed` label. It's less about importance and more about missing expertise :point_up:

> - [Help wanted](https://github.com/mumble-voip/mumble/labels/help%20wanted)

Author: Ok, in that case it might be worth to explain the labels and also to label more issues in that sense.

- [Feature Request](https://github.com/mumble-voip/mumble/labels/feature-request)

> <!-- mention priority labels, e.g. [P2- Critical](https://github.com/mumble-voip/mumble/labels/priority%2FP2%20-%20Important) -->

Reviewer: We don't really use these anymore, so I don't think it's worth mentioning them here

Author: And what is used instead?

Reviewer: Nothing. Since we don't employ any devs that we can tell what to work on, everyone just works on whatever 8s)he wants to. Thus it doesn't make too much sense to have these labels as work is not performed based on them anyways.

Author: But there are still things that are more important than others.

> Under [Milestones](https://github.com/mumble-voip/mumble/milestones) we list the Issues that are planned to be worked on for the next releases.

Reviewer: ```suggestion
```
No, we no longer do that. There still exists one for 1.4.0 but it's not up-to-date (we should probably remove it)

Author: Also not used anymore and what is used instead?

Reviewer: Also nothing. The argumentation is basically the same as above.
Furthermore we intend to switch to a more regular release cycle after 1.4.0 has been released. Thus we won't make a release depend on certain things being implemented or not but simply release what's there at that point instead.

Author: I would still label things that the team thinks are important.
So contributers know what is more useful etc.

Also it could be labeled what issues you agree or disagree with, so that no one tries to implement something that you will not accept in the end.

*Tips:*

> - To avoid unnecessary work and duplication, comment on Issues and inform others what you plan to work on.

Reviewer: ```suggestion
- To avoid unnecessary work and duplication, comment on issues and inform others what you plan to work on.
```

<!-- old:

We are always looking for helping hands willing to contribute to mumble. You'll need a good understanding of C++ and previous knowledge in Qt is definitely helpful. The best way to get into mumble development is to link up with us on {{< wiki IRC />}} ([#mumble-dev on freenode](irc://chat.freenode.net/mumble-dev)), there is always small stuff to do to get you started. Also check the {{< wiki Development "Development wiki page" />}} on more info about the source code and building it.

-->

### Additional projects

#### 3rd party Interfaces

<!-- mention gRPC? -->

> Web interfaces, web-viewers, administration programs, CMS website modules – all neat things we would love to see! These could be written in any language supported by our Ice based RPC Interface. You could start your own project or link up with an {{< wiki "3rd Party Applications" "already existing one" />}}. To get an idea about how interaction with Murmur might look like you can take a look at the example scripts in our git repositories script folder.

Reviewer: ```suggestion
Web interfaces, web-viewers, administration programs, CMS website modules – all neat things we would love to see! These could be written in any language supported by our Ice based RPC Interface. You could start your own project or link up with an {{< wiki "3rd Party Applications" "already existing one" />}}. To get an idea about how interaction with the Mumble server (Murmur) might look like you can take a look at the example scripts in our git repositories script folder.
```

#### Ice authenticators

See {{< wiki Authenticators />}}. We’re happy for every authenticator and authenticator improvement you and the community can bring us!

#### Positional Audio

Most of our positional audio plugins regularly break on game updates. To read on how you can help us update the plugins see the {{< wiki Pluginguide />}} page. You can even fix the update-issue forever for open source games, or if you want to add positional audio to your game, see {{< wiki Link />}}.

> <!-- mention third-party-projects? -->

Reviewer: Do you mean things like e.g. grumble?

Author: Well, strictly seen grumble is an internal project or not?
But still it could be mentioned as well.

There are a lot of projects: pymumble, Webrtc bridge, Bots etc.

Reviewer: Yeah strictly speaking that's true. As nobody from the team is working on it though I consider it to be pretty much 3rdParty by now ^^

Imo we can leave 3rdParty apps out here as this page is supposed to be about contributing to the Mumble project. From my point of view that refers to the "official" components only.
I think we should have a separate page for 3rdParty applications with links to their docs. If someone wants to contribute there, they can find the necessary information at that project's website :point_up:

Author: I would have only mentioned them, like "hey if you are interested in more projects, here are some projects related to Mumble".

Reviewer: In that case I think we can simply link to the 3rdParty apps page from the contribute page for this purpose.

Author: Thats the idea.

Author: I just noticed that 3rd party apps are mentioned in `3rd party Interfaces`...
I think it's not the same, also because it mentions a focus on Ice...

I will add small part mentioning them with a link.

## Non-Coding

### Translation

Mumble currently (April 2020) supports over 15 different languages (38 if including unfinished translations). As more people get to know and use Mumble, versions in their native languages become more and more important.

We are thankful for any contribution on translating. Please check the {{< wiki "Language Translation" "Language Translation wiki page" />}} for more information.

### Testing

To find bugs and polish new releases we need testers.

If you want to test newest development versions, you have two options:

- We provide development snapshots. You can download them from our [Downloads page]({{< relref "downloads" >}}).
<!-- link to exact point? https://www.mumble.info/downloads/#development-snapshots -->

- Build your own version of Mumble, see [Building Mumble](https://github.com/mumble-voip/mumble#building).

If you find any issues, take a look at [Report Bugs](#report-bugs-and-feature-requests).

<!-- mention documentation? Potential problem with TOC linking as the categories have the same name -->
<!-- 
### Documentation

See [below](#documentation). -->

### Mumble theme

<!-- Add more details, descriptions etc.? -->

> You find the Mumble theme in it's own [repository on Github](https://github.com/mumble-voip/mumble-theme).

Reviewer: ```suggestion
You find the Mumble theme in its own [repository on GitHub](https://github.com/mumble-voip/mumble-theme).
```

> #### Skins

Reviewer: Isn't skinning and theming basically the same thing? :thinking:

Author: Kind of.
We could discuss a general renaming.

Author: I opened a discussion: https://github.com/mumble-voip/mumble/discussions/4647

Author: Renamed it, but not in the wiki.

Mumble skins bring variety into Mumbles design and choices for different tastes.

Create new skins for Mumble! Information can be found on our {{< wiki Skinning "Skinning wiki page" />}} and existing skins we know about on the {{< wiki Skins "Skins wiki page" />}}.

# Documentation & Website

Improve our Documentation and Website.

## Documentation

<!-- Todo: 

Add a concrete plan:

> - what documentation we want to have

Reviewer: I guess pretty much all that we can get :eyes:

Author: See below.

> - what should be converted from the wiki

Reviewer: Everything as ultimately we want to get rid of the wiki entirely

Author: Well, that is not really a good answer, as I mentioned I would like to have plan and that includes:
1. Must have category
2. Good to have category
3. Obsolete or can/should be deleted category
4. What is outdated and needs work or what is good and can be converted as is

Reviewer: Given that I have not really read through the available documentation on the wiki, I can't really make this categorization. Ultimately I think non-dev users are best suited to make these decisions as they don't have knowledge of the internals anyways and are thus indeed dependent on the docs

Author: I opened a discussion: #125

> - what is done & what is not

Reviewer: This again is very prone to become outdated...

Author: Thats what Issues are there for.
We can have a todo list, that gets updated.
In this text, we only link to the issue.

Reviewer: I think we should rather use issues and milestones/projects in the GitHub repo directly. And here we could link to the general milestone/project but not to specific issues :thinking:

Author: Sounds good.

- add issues and open pull requests

-->

> We plan to move all our documentation to the Website and Github Repo respectively.

Reviewer: ```suggestion
We plan to move all our documentation to the Website (enduser documentation) and GitHub repo (developer documentation) respectively.
```

Author: Well, essentially that is not correct imo.
1. All documentation is moved to github, because the website is on github. (I know nitpicking :stuck_out_tongue_winking_eye: ).
2. As far as I have seen and understood, we want:
     - enduser documentation on github and website
     - developer documentation on github and website

Reviewer: 1. Fair enough ^^
2. We certainly want the enduser docs on the website. The developer docs idk. Personally I think it's fine to have the developer info only in the main repo. The important keyword here is _main repo_ though. Dev docs should be were the devs are and that's the source code. This makes putting it on the website a bit tedious though (I think) as it's in the wrong repo

Author: I think all of it can (and should) be on the website as well and the discussion between @Kissaki and @davidebeatrici in #35 indicates they want to implement a development documentation as well.

A completely different topic is the question of the repo.
Right now we have most of it in the main repo (Mumble) and it could stay this way (because it can "simply" be converted and implemented on the website), but there are two arguments against it:
If we would have a seperate documentation repo, we could:
- keep the download size and visible structure minimal (though there are methods like git sparse-checkout for it, but that might not be known to everyone)
- keep better track of issues and pull requests specific to documentation

Reviewer: > Dev docs should be were the devs are and that's the source code.

This is really important to me. I don't mind putting the docs on the website _in addition_ to that as long as we find a solution that does not require (manual) duplication of the docs.
I could imagine a process in which the CI in this repo pulls the main repo and includes the developer docs in this way. Alternatively we could probably use something like GitHub Actions to automatically update a mirror of the dev docs in this repo. Thus the dev docs here would be automatically generated from the actual docs in the main repo.

Author: Sounds good to me.

<!-- 

**Todo:**

- Convert parts of the Wiki into Markdown files
- Rewrite parts of the documentation, so it is more recent, comprehensive and fits the style <!-- of the website, markdown files on github -->

Right now you find most of Documentation in our
[wiki](https://wiki.mumble.info).
Any contributions are welcome, improving existing pages or creating new ones with useful content. Also see the {{< wiki "Category:Please_Improve" "Please Improve Category" />}}.

The new places for documentation will be here:

> - [Mumble Github Repo - Docs](https://github.com/mumble-voip/mumble/tree/master/docs)

Reviewer: ```suggestion
- [Mumble GitHub repo - Docs](https://github.com/mumble-voip/mumble/tree/master/docs) for developer-related documentation
```

Author: Regarding my pull request for user-guides that is undecided for now.
But if you decide it should be solely on the website...

Reviewer: This is actually something that we have not (yet) decided as a team yet. That's solely what I think would make the most sense.
Maybe @davidebeatrici and @Kissaki can given their opinion on this as well?

Regarding your pull request: Given that these are Markdown files anyways I don't think it matters all that much as to in which repo it gets initially created as it can be easily migrated if needed.

> - [Website's Github repo - Documentation](https://github.com/mumble-voip/mumble-www/tree/master/hugo/content/documentation)

Reviewer: ```suggestion
- [Website's GitHub repo - Documentation](https://github.com/mumble-voip/mumble-www/tree/master/hugo/content/documentation) for enduser-related documentation
```

Author: see above

#### Documentation Issues & Pull requests

> - [Documentation Issues on Github](https://github.com/mumble-voip/mumble/labels/documentation)

Reviewer: ```suggestion
- [Documentation issues on GitHub](https://github.com/mumble-voip/mumble/labels/documentation)
```

<!--
Pull requests and issues:

- [DOCS: Add Userguides to the documentation](https://github.com/mumble-voip/mumble/pull/4628) 

- (Work-in-Progress) [C++ Guidelines & tools](https://github.com/mumble-voip/mumble/discussions/4261)

-->

---

<!-- 
**Ice methods**: 
<!-- outdated? -->
<!-- Is this here now: https://www.mumble.info/documentation/slice/1.3.0/html/Murmur.html  ? -->
<!--
We have a [generated documentation](http://mumble.sourceforge.net/slice/) for exposed Ice methods. Although probably only developers using the doc will be able to understand it, please give us a note on any inconsistencies, errors or needed clarifications if you find them.
-->

### External Documentation

- **Wikipedia**: Someone that is not a development team member needs to maintain the [article(s) at Wikipedia](https://en.wikipedia.org/wiki/Mumble_%28software%29).

## Website

<!-- 
Todo:
Add more instructions?
Here or at the Mumble-www repo.

Such as:

> -Writing guidelines (markdown etc.)

Reviewer: That's probably useful to have in the repo :+1:

Author: I agree.

> -Content (what is planned and needs improvement etc.?) 

Reviewer: For this we should use issues in the www repo :point_up:

Author: I agree.

-->

> This website has a Github repository at [mumble-www](https://github.com/mumble-voip/mumble-www).

Reviewer: ```suggestion
This website has a GitHub repository at [mumble-www](https://github.com/mumble-voip/mumble-www).
```

It consists of two parts:

> - a webserver written in GO

Reviewer: Is that even true? I always thought that hugo was only generating the static HTML pages that are then uploaded to the server :thinking:

Author: That is true, it's like a chain.
Still the webserver serves html, while hugo converts the markdown stuff and some other things into html.
Thats at least how I understand it.

Author: I guess we should include all the description and explanation into the readme.md of mumble-www.
So in the end I will just mention some specific links.

> - Most of the content is build with [hugo](https://gohugo.io/) and written in **Markdown**, **HTML** and some css.

Reviewer: ```suggestion
- Most of the content is built with [hugo](https://gohugo.io/) and written in **Markdown**, **HTML** and some css.
```

You can:

- report [issues](https://github.com/mumble-voip/mumble-www/issues)
- create [requests](https://github.com/mumble-voip/mumble-www/issues), 
- create pull requests for changes to the website

## Community

Be a part of our community and help others!

> * [Discussions on Github](https://github.com/mumble-voip/mumble/discussions)

Reviewer: ```suggestion
* [Discussions on GitHub](https://github.com/mumble-voip/mumble/discussions)
```

* Our [issue tracker](https://github.com/mumble-voip/mumble/issues)
* IRC: [#mumble on irc.freenode.net](irc://irc.freenode.org/mumble)
* Chat: [Mumble room on Riot/Matrix](https://riot.im/app/#/room/#mumble:matrix.org) (shares messages with IRC)
* Our [Forums](https://forums.mumble.info/)

> <!-- forums deprecated? -->

Reviewer: The forums have just received an update to a brand new forums software. So no: not deprecated :)

Author: Sounded different some time ago. But ok, I will delete the comment sometime.

# Hosting

Host your own Mumble Server and register it on our Public Server List.

> <!-- add server guide {{< wiki Running_Murmur /v}} -->

Reviewer: ```suggestion
<!-- add server guide {{< wiki Running_Murmur />}} -->
```
I think this should fix the CI

Author: So this little problem lead to total failure :smile: .

Reviewer: Apparently - it's especially weird since it's in a comment xD

The best way to get into Mumble development is to link up with us on IRC ([#mumble-dev on freenode](irc://chat.freenode.net/mumble-dev)), also available via Matrix (Join: `#freenode_#mumble-dev:matrix.org` or use [Direct Link via Element.io (Web App)](#freenode_#mumble-dev:matrix.org)).

> Alternatively you can create or comment (on) a Pull request or issue report.

Reviewer: ```suggestion
Alternatively you can create or comment (on) a pull request or issue report.
```

#### 3rd party Interfaces

> Web interfaces, web-viewers, administration programs, CMS website modules – all neat things we would love to see! These could be written in any language supported by our Ice based or gRPC based RPC Interfaces. You could start your own project or link up with an {{< wiki "3rd Party Applications" "already existing one" />}}. To get an idea about how interaction with the Mumble server might look like, you can take a look at the example scripts in our [GitHub repository script folder](https://github.com/mumble-voip/mumble/tree/master/scripts).

Reviewer: Uhm I don't think that a lot of scripts there are about this purpose. What this text is referring to is probably https://github.com/mumble-voip/mumble-scripts, isn't it? :thinking:

Author: Ah, good to know.

### Testing

To find bugs and polish new releases (before and after release) we need testers.

If you want to test the newest development versions, you have two options:

- We provide development snapshots. You can download them from our [Downloads page]({{< relref "downloads" >}}).
<!-- link to exact point? https://www.mumble.info/downloads/#development-snapshots -->

> - Build the latest master of Mumble, see [Building Mumble](https://github.com/mumble-voip/mumble#building).

Reviewer: ```suggestion
- Build the latest version of Mumble (`master` branch on GitHub), see [Building Mumble](https://github.com/mumble-voip/mumble#building).
```

Author: Well I tried to avoid `latest version`, as that might be confusing.

What about `latest development version`?

Reviewer: Sounds good to me :+1:

Author: What about `latest development state`?
version sounds like a release version...

I was also thinking about adding information to the building instructions about how to use specific versions (it's obvious for people who know git, but for beginners it might be useful).
